Bookbinder HWP binding

jlashner opened the pull request:
This PR adds functionality to the bookbinder which adds processed HWP data to books.

If the `hwp_root` is passed, this will load in the processed HWP g3 files, and use it to fill out the ancillary data files in books.

This tries to replicate the format described in the tod_format doc as closely as possible. For the negative integer values that report HWP conditions, I use a bitmask with flags for when the hwp is stationary, stable, etc. based on what's in the processed data file. I assume this is something people have opinions about so we should probably discuss the correct flags for this.

This makes a few other small changes in the bookbinder that I came across when testing:
- the `ancil` field is now added to ancillary frames even if there's no data (Matthew also noticed this in #388)
- G3Files are now properly closed with EndProcessing frames

I've tested this on a few books on simons1, but probably needs to be tested with the whole p10r2 archive. What is the best way to do this?

algadoreo commented on sotodlib/io/bookbinder.py:
```diff
@@ -571,6 +582,37 @@ def flush(self, flush_time=None):
+
+        ## Adds HWP data if present
+        if self.hwp_loader is not None:
+            if self.hwp_loader.hwp_angle_interp is not None:
+                ts = np.array(smurf_data.times) / core.G3Units.s
+                hwp_angle = self.hwp_loader.hwp_angle_interp(ts)
+                hwp_angle = np.mod(hwp_angle, 2*np.pi)
+
+                hwp_locked = self.hwp_loader.locked_interp(ts)
+                hwp_stable = self.hwp_loader.stable_interp(ts)
+                hwp_rate = self.hwp_loader.stable_interp(ts)
+
+                flags = np.zeros_like(hwp_angle, dtype=np.int32)
+                m = hwp_rate < MIN_HWP_RATE
+                flags[m] = np.bitwise_or(flags[m], HWPFlags.STATIONARY)
+                m = hwp_locked != 1
+                flags[m] = np.bitwise_or(flags[m], HWPFlags.UNLOCKED)
+                m = hwp_stable != 1
+                flags[m] = np.bitwise_or(flags[m], HWPFlags.UNSTABLE)
+                m = np.isnan(hwp_angle)
+                flags[m] = np.bitwise_or(flags[m], HWPFlags.NO_DATA)
+
+                hwp_angle[flags != 0] = -flags[flags != 0]
+            else:
+                hwp_angle = np.full(len(smurf_data.times), -HWPFlags.NO_DATA)
+
```
Perhaps this should be in its own method (e.g., `get_hwp_angles`), to avoid having too much in `flush`.

jlashner replied:
Sounds good, I can add this in a method of the HWPLoader

algadoreo commented on sotodlib/io/bookbinder.py:
```diff
@@ -1034,6 +1090,14 @@ def compile_mfile_dict(self):
+
+        if self.hwp_loader is not None:
+            if self.hwp_loader.hwp_angle_interp is not None:  # If there's data
+                hwp_rate = float(np.mean(self.hwp_loader.data['slow']['hwp_rate']))
```
What’s the reason for the `float()` here?

jlashner replied:
I had to add this because mean returns a `np.float`, which isn't formatted well in yaml

algadoreo replied:
In that case, could you add a brief comment for future viewers?

algadoreo commented on sotodlib/io/bookbinder.py:
```diff
@@ -1125,3 +1196,105 @@ def __call__(self):
+                continue
+
+            for bname, block in zip(frame['block_names'], frame['blocks']):
+                if bname not in data:
+                    data[bname] = {
+                        k: [] for k in block.keys()
+                    }
+                    times[bname] = []
+
+                ts = np.array(block.times) / core.G3Units.s
+
+                # Includes a sec before and after for interpolation
+                m = ((start - 1) < ts) & (ts < (stop + 1))
+
+                for field, d in block.items():
+                    data[bname][field].append(np.array(d)[m])
+                times[bname].append(ts[m])
+
+        for bname in data:
+            for field in data[bname]:
+                data[bname][field] = np.hstack(data[bname][field])
+            times[bname] = np.hstack(times[bname])
+
+        self.data = data
+        self.times = times
+        if len(data) != 0:
+            self.hwp_angle_interp = interp1d(
+                times['fast'], np.unwrap(data['fast']['hwp_angle']), 
+                bounds_error=False)
+            self.locked_interp = interp1d(
+                times['slow'], data['slow']['locked'], bounds_error=False)
+            self.stable_interp = interp1d(
+                times['slow'], data['slow']['stable'], bounds_error=False)
+            self.hwp_rate_interp = interp1d(
+                times['slow'], data['slow']['hwp_rate'], bounds_error=False)
+        else:
+            self.hwp_angle_interp = None
+            self.locked_interp = None
+            self.stable_interp = None
+            self.hwp_rate_interp = None
```
Do we anticipate needing this elsewhere? If so, might be good as a separate module/file in the sotodlib/io/ directory.

I usually put a newline at the end of my files out of habit, but I guess that’s not strictly necessary.

jlashner replied:
Ah I'm too spoiled by the autoformatter that Brian added in ocs/socs haha. I'll add it. I'm not sure how useful it is elsewhere in sotodlib, so I kind of think we should just keep it here for now since we can easily move it to a separate module later if that ends up being helpful

algadoreo reviewed (approved):
This looks good to me on a read-through. Since I don’t have anything to test it with, do you plan on adding a unit test to test_bookbinder.py? (Doesn’t have to be now.) Otherwise, my only comments are on the organizational aspects.

Also, thanks for adding the `EndProcessing` frame—I wasn’t aware that was a thing.

guanyilun commented on sotodlib/io/bookbinder.py:
```diff
@@ -571,6 +582,37 @@ def flush(self, flush_time=None):
+
+                hwp_locked = self.hwp_loader.locked_interp(ts)
+                hwp_stable = self.hwp_loader.stable_interp(ts)
+                hwp_rate = self.hwp_loader.stable_interp(ts)
```
typo? hwp_rate_interp?

jlashner replied:
Yep, typo, good catch

guanyilun commented on sotodlib/io/imprinter.py:
```diff
@@ -257,7 +257,7 @@ def register_books(self, bids, obs_lists, commit=True, session=None):
         if commit: session.commit()
 
-    def bind_book(self, book, session=None, output_root="out", message=""):
+    def bind_book(self, book, session=None, output_root="out", message="", hwp_root=None):
```
I think I might prefer to no change to the function signature here because they need to be propagate downstream. Shouldn't hwp_root be already specified in the g3tsmurf config file which is stored in `self.sources[book.tel_tube]['g3tsmurf']`? This will be easier for downstream tasks.

guanyilun commented on sotodlib/io/bookbinder.py:
```diff
@@ -1125,3 +1196,105 @@ def __call__(self):
+            except:
+                continue
+
+            if not (((start//1e5) - 1) <= tcode <= stop//1e5):
```
Why not stop//1e5 + 1?

jlashner replied:
My reasoning for these start time buffers is just because the sample with timestamp `t` will fall in a file timestamped before the time `t`. So we have to check the file before the start time, but the file after the end time shouldn't contain any samples that we care about.

However as you mention below, the timestamp in the filename may not actually be a lower bound for the samples in file, so you are right that we should probably include a buffer on the `stop` time also.

guanyilun commented on sotodlib/io/bookbinder.py:
```diff
@@ -1125,3 +1196,105 @@ def __call__(self):
+            self.files = []
+            return self.files
+
+        # Add file before for good measure
```
What about the file afterward?

guanyilun commented on sotodlib/io/bookbinder.py:
```diff
@@ -1125,3 +1196,105 @@ def __call__(self):
+                ts = np.array(block.times) / core.G3Units.s
+
+                # Includes a sec before and after for interpolation
+                m = ((start - 1) < ts) & (ts < (stop + 1))
```
We should be careful about the assumption here. The start and stop times in bookbinder come from g3tsmurf db. @kmharrington 
could clarify how that timestamp is obtained. These timestamps often do not line up with the actual .times field, probably more so when timing system is on. Here it's assuming they can't be off by more than 1s. Is that a safe assumption?

jlashner replied:
Ah this is a really good point. Since there is this uncertainty, I think we may as well just use a much larger buffer, of like `30 sec` or so on each side.

jlashner commented:
Great thanks for your comments! Just added a commit that addresses both of your feedback, so if you could re-review that would be awesome.

As far as testing goes, I could add unit tests for the HWPLoader class. I think one thing we're lacking is a end-to-end test of the bookbinder. One that generates a small fake data archive and attempts to run the `bind_book` function, which should cover a lot of the bookbinder class. Maybe we can add that in a separate PR?

guanyilun reviewed (approved):
looking good, thanks for the changes!

guanyilun commented:
> I think one thing we're lacking is a end-to-end test of the bookbinder. One that generates a small fake data archive and attempts to run the `bind_book` function, which should cover a lot of the bookbinder class. Maybe we can add that in a separate PR?

+1 for setting end-to-end testing for bookbinder

algadoreo commented:
I’m good with merging this. End-to-end testing is definitely outside the scope of this PR. If we’re going to make one though, we should include the other HK stuff since that has not been tested at all.

kmharrington reviewed (changes requested):
Two in-line comments then another question:

I see HWP rate in the `M_index.yaml`. Can we add an overall flag there as well. Something that we could build a database off of later for indexing what the HWP is doing. I think that it's basically going to be some level of all or any calls to the timestream flag you built. (What those flags mean also need to go into the tod2maps docs).

jlashner commented on sotodlib/io/bookbinder.py:
```diff
@@ -6,6 +6,16 @@
 import yaml
+from scipy.interpolate import interp1d
+# 
+class HWPFlags:
```
This doesn't add an additional field for flags. In the tod_format docs it says that the `hwp_enc` field should be a float between [0, 2pi), and a negative integer to report any other conditions. So these flags are just used to determine what that negative integer should be.

I actually wasn't sure if these are actually good flags to use and wanted to open it up to a larger discussion, since I'm not so familiar with the HWP data and don't really know what goes into constructing the angle (to be honest I don't really know how "locked" and "stable" should be interpreted). Do you have any suggestions for what flags would be useful here?

jlashner commented:
Hi all, thanks for the great feedback! I was finally able to finish implementing changes.

I changed the HWPFlags based on my new understanding of what's in the level 2.5 HWP files, so we only are flagging if "stationary" (i.e. stable and not locked) and "no data" (i.e. not stable).

I also added the HWP rate taken once every 10 minutes to the meta file, along with a description to tell us whether the HWP is stationary, changing, or constant.

Let me know if there are any other issues!

kmharrington commented:
looks good to me. Sorry for the delay on the review

kmharrington reviewed (approved):
Looks good to me, sorry for the detail on the review.

jlashner commented:
Should we bother merging this since we decided to not include HWP data in the ancillary data files? I'm fine with just closing it.

kmharrington commented:
If the changes aren't breaking changes then I think I'd like to "keep this in our back pocket" so that if the plan changes again no one has to re-do all you've done.
